src/coreclr/jit/fgdiagnostic.cpp

```diff
@@ -2726,6 +2726,9 @@ bool BBPredsChecker::CheckEhTryDsc(BasicBlock* block, BasicBlock* blockPred, EHb
         return true;
     }
 
+    // Async resumptions are allowed to jump into try blocks at any point. They
+    // are introduced late enough that the invariant of single entry is no
+    // longer necessary.
```

Comment:
Roslyn rewrites awaits in `try` blocks as a switch at the single entry that dispatches to appropriate resume location. 
It is nice that runtime async can resume directly into `try` blocks.

Awaits in handlers might be harder to allow, I guess.

Reply:
> Roslyn rewrites awaits in `try` blocks as a switch at the single entry that dispatches to appropriate resume location. Nested cases get dispatched multiple times. It is nice that runtime async can resume directly into `try` blocks, so we can leave awaits in trys as-is.

Yeah, this definitely simplifies things. For OSR we have to do the similar transformation as Roslyn, creating the "step" blocks at the beginning of try's.

> Awaits in handlers might be harder to allow, I guess.

Indeed, seems like it would be very hard, so I'm glad we disallow that.